File: src/services/inference/pure_dep_graph_operations.ml
(*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

open Utils_js

(* `rdep_closure graph files` returns all files in `graph` which are reachable
 * from `files`, directly or indirectly. *)
let rdep_closure =
  let rec loop graph files acc =
    FilenameSet.fold
      (fun file acc ->
        let entry = FilenameGraph.find_backward_opt file graph in
        match entry with
        | Some files ->
          let files = FilenameSet.diff files acc in
          let acc = FilenameSet.union files acc in
          loop graph files acc
        | None -> acc)
      files
      acc
  in
  (fun graph files -> loop graph files files)

(* `calc_direct_dependencies graph files` will return the set of direct dependencies of
   `files`. This set includes `files`. *)
let calc_direct_dependencies dependency_graph files =
  FilenameSet.fold
    (fun file acc ->
      match FilenameGraph.find_opt file dependency_graph with
      | Some files -> FilenameSet.union files acc
      | None -> acc)
    files
    files

(** [calc_direct_dependents graph files] will return the set of direct dependents of
  [files]. This set includes [files]. *)
let calc_direct_dependents dependency_graph files =
  FilenameSet.fold
    (fun file acc ->
      match FilenameGraph.find_backward_opt file dependency_graph with
      | Some files -> FilenameSet.union files acc
      | None -> acc)
    files
    files

(** [calc_all_dependents graph files] will return the set of direct and transitive dependents of
  [files]. This set include [files].

  A file is a dependent of [files] whenever its code depends on any file whose *signature*, in
  turn, directly or transitively depends on [files]. *)
let calc_all_dependents ~sig_dependency_graph ~implementation_dependency_graph files =
  let all_sig_dependents = rdep_closure sig_dependency_graph files in
  let all_dependents = calc_direct_dependents implementation_dependency_graph all_sig_dependents in
  (all_sig_dependents, all_dependents)

File: rust_port/crates/flow_services_inference/src/pure_dep_graph_operations.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

//! Contains pure functions which perform calculations on the dependency graph

use dupe::Dupe;
use flow_common_utils::graph::GraphLike;
use flow_data_structure_wrapper::ord_set::FlowOrdSet;
use flow_parser::file_key::FileKey;

/// `rdep_closure graph files` returns all files in `graph` which are reachable
/// from `files`, directly or indirectly.
fn rdep_closure<G>(graph: &G, files: &FlowOrdSet<FileKey>) -> FlowOrdSet<FileKey>
where
    G: GraphLike<FileKey> + ?Sized,
{
    fn loop_impl<G>(graph: &G, files: &FlowOrdSet<FileKey>, acc: &mut FlowOrdSet<FileKey>)
    where
        G: GraphLike<FileKey> + ?Sized,
    {
        for file in files {
            if let Some(rdeps) = graph.find_backward_opt(file) {
                let mut new_files = FlowOrdSet::new();
                for rdep in rdeps {
                    if !acc.contains(rdep) {
                        new_files.insert(rdep.dupe());
                    }
                }
                if !new_files.is_empty() {
                    for f in &new_files {
                        acc.insert(f.dupe());
                    }
                    loop_impl(graph, &new_files, acc);
                }
            }
        }
    }

    let mut result = files.dupe();
    loop_impl(graph, files, &mut result);
    result
}

/// `calc_direct_dependencies graph files` will return the set of direct dependencies of
/// `files`. This set includes `files`.
pub fn calc_direct_dependencies(
    dependency_graph: &(impl GraphLike<FileKey> + ?Sized),
    files: &FlowOrdSet<FileKey>,
) -> FlowOrdSet<FileKey> {
    let mut result = files.dupe();

    for file in files {
        if let Some(deps) = dependency_graph.find_opt(file) {
            for dep in deps {
                result.insert(dep.dupe());
            }
        }
    }

    result
}

/// `calc_direct_dependents graph files` will return the set of direct dependents of
/// `files`. This set includes `files`.
pub fn calc_direct_dependents(
    dependency_graph: &(impl GraphLike<FileKey> + ?Sized),
    files: &FlowOrdSet<FileKey>,
) -> FlowOrdSet<FileKey> {
    let mut result = files.dupe();

    for file in files {
        if let Some(deps) = dependency_graph.find_backward_opt(file) {
            for dep in deps {
                result.insert(dep.dupe());
            }
        }
    }

    result
}

/// `calc_all_dependents graph files` will return the set of direct and transitive dependents of
/// `files`. This set include `files`.
///
/// A file is a dependent of `files` whenever its code depends on any file whose *signature*, in
/// turn, directly or transitively depends on `files`.
pub fn calc_all_dependents(
    sig_dependency_graph: &(impl GraphLike<FileKey> + ?Sized),
    implementation_dependency_graph: &(impl GraphLike<FileKey> + ?Sized),
    files: &FlowOrdSet<FileKey>,
) -> FlowOrdSet<FileKey> {
    let all_sig_dependents = rdep_closure(sig_dependency_graph, files);
    calc_direct_dependents(implementation_dependency_graph, &all_sig_dependents)
}
